### PR TITLE
Returning bytesRead as part of searchResult

### PR DESCRIPTION
**Pull request by Thejas-bhat**
Refactoring the bytesRead accounting to now be accounted at a per query level. The searchResult returned by SearchInContext now has an extra field called bytesRead which helps to track the cost associated with the query. 

The implementation to achieve this basically involves passing a context.Context which has a callback that is used to track or update the bytes read associated with some operation. This change of adding an extra parameter context to the searcher is an invasive one, and hence involves updating a lot of existing files to accommodate the same. However, the good thing is that the adding of context to searcher is in a way a future proof change, which hopefully might be useful to do any more useful things down at the searcher level and relay them back to the indexReader level.

### CODE REVIEW

File: document/document.go [hunk 1]
```diff
@@ -30,9 +30,18 @@ func init() {
...
-	id              string  `json:"id"`
-	Fields          []Field `json:"fields"`
-	CompositeFields []*CompositeField
+	id                string  `json:"id"`
+	Fields            []Field `json:"fields"`
+	CompositeFields   []*CompositeField
+	storedFieldsBytes uint64
+}
+
+func (d *Document) SetStoredFieldsBytes(val uint64) {
```
Comment (abhinavdangeti) [inline document/document.go:39]: Drop this method you can updated storedFieldsBytes by simply exporting it (like CompositeFields above).

File: go.mod [hunk 2]
```diff
@@ -40,3 +40,17 @@ require (
 	github.com/spf13/pflag v1.0.5 // indirect
 	golang.org/x/sys v0.0.0-20220520151302-bc2c85ada10a // indirect
 )
+
+replace github.com/blevesearch/zapx/v15 => ../zapx
```
Comment (abhinavdangeti) [inline go.mod:44]: Remove all these replaces.

File: index/scorch/snapshot_index.go [hunk 3]
```diff
@@ -430,13 +432,17 @@ func (i *IndexSnapshot) Document(id string) (rv index.Document, err error) {
 	segmentIndex, localDocNum := i.segmentIndexAndLocalDocNumFromGlobal(docNum)
 
 	rvd := document.NewDocument(id)
-	prevBytesRead := i.segment[segmentIndex].segment.BytesRead()
+	var totalStoredBytes int
```
Comment (abhinavdangeti) [inline index/scorch/snapshot_index.go:435]: remove variable.

File: index/scorch/snapshot_index.go [hunk 4]
```diff
@@ -430,13 +432,17 @@ func (i *IndexSnapshot) Document(id string) (rv index.Document, err error) {
...
 
 	err = i.segment[segmentIndex].VisitDocument(localDocNum, func(name string, typ byte, val []byte, pos []uint64) bool {
 		if name == "_id" {
 			return true
 		}
 
+		// track uncompressed stored fields bytes as part of IO stats.
+		// However, ideally we'd need to track the compressed on-disk value
+		// Keeping that TODO for now until we have a cleaner way.
+		totalStoredBytes += len(val)
```
Comment (abhinavdangeti) [inline index/scorch/snapshot_index.go:445]: rvd.StoredFieldsBytes += len(val)

Separate question - shouldn't you be doing this only when the field is actually "stored"?

File: index/upsidedown/index_reader.go [hunk 5]
```diff
@@ -15,11 +15,12 @@
 package upsidedown
 
 import (
+	"context"
 	"reflect"
 
 	"github.com/blevesearch/bleve/v2/document"
 	index "github.com/blevesearch/bleve_index_api"
-	"github.com/blevesearch/upsidedown_store_api"
+	store "github.com/blevesearch/upsidedown_store_api"
```
Comment (abhinavdangeti) [inline index/upsidedown/index_reader.go:23]: This changes seems unrelated?

File: index/upsidedown/reader.go [hunk 6]
```diff
@@ -22,7 +22,7 @@ import (
 
 	"github.com/blevesearch/bleve/v2/size"
 	index "github.com/blevesearch/bleve_index_api"
-	"github.com/blevesearch/upsidedown_store_api"
+	store "github.com/blevesearch/upsidedown_store_api"
```
Comment (abhinavdangeti) [inline index/upsidedown/reader.go:25]: Unrelated change?

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]
Mostly nits, I'll make another pass over this soon.
Meanwhile you should get some perf results with these changes.

File: document/document.go [hunk 7]
```diff
@@ -30,9 +30,18 @@ func init() {
...
+	d.storedFieldsBytes = val
+}
+
+func (d *Document) GetStoredFieldsBytes() uint64 {
```
Comment (abhinavdangeti) [inline document/document.go:43]: This method can go away too, with storedFieldsBytes exported.

File: index/upsidedown/index_reader.go (hunk 5 above)
Comment (Thejas-bhat) [reply index/upsidedown/index_reader.go:23]: This popped up because of the vscode extension I believe, removing it.

File: index/upsidedown/reader.go (hunk 6 above)
Comment (Thejas-bhat) [reply index/upsidedown/reader.go:25]: This popped up because of the vscode extension I believe, removing it.

File: document/document.go (hunk 1 above)
Comment (Thejas-bhat) [reply document/document.go:39]: Done.

File: document/document.go (hunk 7 above)
Comment (Thejas-bhat) [reply document/document.go:43]: This Get type of API might be necessary since the struct is hidden behind the Document interface.

File: go.mod (hunk 2 above)
Comment (Thejas-bhat) [reply go.mod:44]: Yeah just got pushed to the review as part of "git add .". Removing them.

File: index/scorch/snapshot_index.go (hunk 4 above)
Comment (Thejas-bhat) [reply index/scorch/snapshot_index.go:445]: This callback is called only for the stored fields. The VisitDocument API ends up calling the VisitStoredFields in the zap layer which loads the bytes pertaining to the stored fields, uncompresses them and iterates over each of these fields and applies this callback.

File: index/scorch/snapshot_index.go (hunk 3 above)
Comment (Thejas-bhat) [reply index/scorch/snapshot_index.go:435]: Done.

**Comment by abhinavdangeti** [comment]
@Thejas-bhat 

- bleve_index_api@v1.0.5
- scorch_segment_api/v2@v2.1.4
- zapx/v11@11.3.7
- zapx/v12@12.3.7
- zapx/v13@13.3.7
- zapx/v14@14.3.7
- zapx/v15@15.3.7

File: index/scorch/snapshot_index.go [hunk 8]
```diff
@@ -37,6 +38,7 @@ import (
 // re usable, threadsafe levenshtein builders
 var lb1, lb2 *lev.LevenshteinAutomatonBuilder
 
+type diskStatsReporter segment.DiskStatsReporter
```
Comment (abhinavdangeti) [inline index/scorch/snapshot_index.go:41]: Is this step necessary, can't we just type assert to segment.DiskStatsReporter?

File: index_impl.go [hunk 9]
```diff
@@ -469,7 +469,20 @@ func (i *indexImpl) SearchInContext(ctx context.Context, req *SearchRequest) (sr
...
 	}()
 
-	searcher, err := req.Query.Searcher(indexReader, i.m, search.SearcherOptions{
+	// This callback and variable handles the tracking of bytes read
+	//  1. as part of creation of tfr and its Next() calls which is
+	//     accounted by invoking this callback when the TFR is closed.
+	//  2. the docvalues portion (accounted in collector) and the retrieval
+	//     of stored fields bytes (by LoadAndHighlightFields)
+	var totalBytesRead uint64
+	SendBytesRead := func(bytesRead uint64) {
```
Comment (abhinavdangeti) [inline index_impl.go:478]: Does this closure need to be exported? Would the compiler complain on sendBytesRead?

**Review by abhinavdangeti — COMMENTED** [review]
Looking good, couple minor comments.

File: index_impl.go (hunk 9 above)
Comment (Thejas-bhat) [reply index_impl.go:478]: Aah, my bad.

File: index/scorch/snapshot_index.go (hunk 8 above)
Comment (Thejas-bhat) [reply index/scorch/snapshot_index.go:41]: Yeah at a couple of places, there are library name and variable name conflicts, for eg. `indexSnapshot.segment.segment `.

Comment (abhinavdangeti) [reply index/scorch/snapshot_index.go:41]: Ah variable and import conflict. Should update the variable name then (it looks quite confusing :)). I'll make that change separately.

**Review by abhinavdangeti — APPROVED** [review]
